File: note_oc/FECollectionViewFlowLayout.m
//
//  FECollectionViewFlowLayout.m
//  laboratory
//
//  Created by hryan on 16/3/11.
//  Copyright © 2016年 fe. All rights reserved.
//

#import "FECollectionViewFlowLayout.h"

@implementation FECollectionViewFlowLayout


const NSInteger kMaxCellSpacing = 15;


-(BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds{
    return YES;
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    NSArray* attributesToReturn = [super layoutAttributesForElementsInRect:rect];
    for (UICollectionViewLayoutAttributes* attributes in attributesToReturn) {
        if (nil == attributes.representedElementKind) {
            NSIndexPath* indexPath = attributes.indexPath;
            attributes.frame = [self layoutAttributesForItemAtIndexPath:indexPath].frame;
        }
    }
    return attributesToReturn;
}

- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
{
    
    
    UICollectionViewLayoutAttributes *currentItemAttributes = [super layoutAttributesForItemAtIndexPath:indexPath];
    
    if (indexPath.section != 0)
    {
        return currentItemAttributes;
    }
    
    UIEdgeInsets sectionInset = [(UICollectionViewFlowLayout *)self.collectionView.collectionViewLayout sectionInset];
    
    if (indexPath.item == 0) { // first item of section
        CGRect frame = currentItemAttributes.frame;
        frame.origin.x = sectionInset.left+20; // first item of the section should always be left aligned
        currentItemAttributes.frame = frame;
        
        return currentItemAttributes;
    }
    
    NSIndexPath* previousIndexPath = [NSIndexPath indexPathForItem:indexPath.item-1 inSection:indexPath.section];
    CGRect previousFrame = [self layoutAttributesForItemAtIndexPath:previousIndexPath].frame;
    CGFloat previousFrameRightPoint = previousFrame.origin.x + previousFrame.size.width + kMaxCellSpacing;
    
    CGRect currentFrame = currentItemAttributes.frame;
    CGRect strecthedCurrentFrame = CGRectMake(0, currentFrame.origin.y, self.collectionView.frame.size.width, currentFrame.size.height);
    
    if ( !CGRectIntersectsRect(previousFrame, strecthedCurrentFrame) )
    {
        CGRect frame = currentItemAttributes.frame;
        frame.origin.x = sectionInset.left;// first item on the line should always be left aligned
        if (frame.origin.x < 20 )
        {
            frame.origin.x = frame.origin.x + 20;
        }
        currentItemAttributes.frame = frame;
        return currentItemAttributes;
    }
    
    CGRect frame = currentItemAttributes.frame;
    frame.origin.x = previousFrameRightPoint;
    currentItemAttributes.frame = frame;
    return currentItemAttributes;
}


@end

File: note_oc/FECollectionViewFlowLayout.swift
//
//  FECollectionViewFlowLayout.swift
//  laboratory
//
//  Created by hryan on 16/3/10.
//  Copyright © 2016年 fe. All rights reserved.
//

import UIKit

let kMaxCellSpacing: CGFloat = 5

class FECollectionViewFlowLayout: UICollectionViewFlowLayout {

    override func layoutAttributesForElementsInRect(rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        
        let attributesToReturn = super.layoutAttributesForElementsInRect(rect)
        
        for attributes in attributesToReturn!{
//            if  nil == attributes.representedElementKind {
//                
//            }
            let indexPath = attributes.indexPath
            attributes.frame = (self.layoutAttributesForItemAtIndexPath(indexPath)?.frame)!;
        }
        
        return attributesToReturn
    }
    
    
    override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
        
       let currentIntemAttributes =  super.layoutAttributesForItemAtIndexPath(indexPath)
        
        if(indexPath.section == 1 || indexPath.section == 2){
            return currentIntemAttributes
        }
        
        let sectionInset = ( self.collectionView?.collectionViewLayout as! UICollectionViewFlowLayout).sectionInset
        if(indexPath.item == 0) {
            var frame = currentIntemAttributes?.frame
            frame?.origin.x = sectionInset.left + 20
            frame?.origin.y = sectionInset.top + 10
            currentIntemAttributes?.frame = frame!
            return currentIntemAttributes
        }
        
        let previousIndexpath = NSIndexPath(forItem: (indexPath.item-1), inSection: indexPath.section)
        
        let previousFrame = super.layoutAttributesForItemAtIndexPath(previousIndexpath)?.frame
        let previousFrameRightPoint = (previousFrame?.size.width)! + kMaxCellSpacing + (previousFrame?.origin.x)!
        let currentFrame = currentIntemAttributes?.frame
        let strecthedCurrentFrame = CGRectMake(0, currentFrame!.origin.y, (self.collectionView?.frame.size.width)!, currentFrame!.size.height)
        
        
        if (!CGRectIntersectsRect(previousFrame!, strecthedCurrentFrame)) {
            var frame = currentIntemAttributes?.frame
            frame?.origin.x = sectionInset.left
            
            if(frame?.origin.x < 10) {
                frame?.origin.x = (frame?.origin.x)! + 10
            }
            currentIntemAttributes?.frame = frame!
            return currentIntemAttributes
        }
        
        var frame = currentIntemAttributes?.frame
        frame?.origin.x = previousFrameRightPoint
//        frame?.origin.y = (previousFrame?.origin.y)!
        currentIntemAttributes?.frame = frame!
        return currentIntemAttributes
        
    }
    
}
